File: src/DataBlock.cpp
//===- DataBlock.cpp -------------------------------------------*- C++ -*-===//
//
//  Copyright (C) 2020 GrammaTech, Inc.
//
//  This code is licensed under the MIT license. See the LICENSE file in the
//  project root for license terms.
//
//  This project is sponsored by the Office of Naval Research, One Liberty
//  Center, 875 N. Randolph Street, Arlington, VA 22203 under contract #
//  N68335-17-C-0700.  The content of the information does not necessarily
//  reflect the position or policy of the Government and no official
//  endorsement should be inferred.
//
//===----------------------------------------------------------------------===//
#include "Serialization.hpp"
#include <gtirb/ByteInterval.hpp>
#include <gtirb/DataBlock.hpp>
#include <gtirb/proto/DataBlock.pb.h>

using namespace gtirb;

void DataBlock::toProtobuf(MessageType* Message) const {
  nodeUUIDToBytes(this, *Message->mutable_uuid());
  Message->set_size(this->Size);
}

DataBlock* DataBlock::fromProtobuf(Context& C, ByteInterval* Parent,
                                   const MessageType& Message) {
  // Because we do not have an offset, we cannot create the data block and
  // set its parent at the same time.
  auto* DO = DataBlock::Create(C, Message.size());
  DO->setByteInterval(Parent);
  if (!setNodeUUIDFromBytes(DO, Message.uuid()))
    return nullptr;
  return DO;
}

uint64_t DataBlock::getOffset() const {
  assert(Parent &&
         "invalid call to DataBlock::getOffset: Parent must not be null!");
  return Parent->nodeToBlock(this).getOffset();
}

std::optional<Addr> DataBlock::getAddress() const {
  if (!Parent) {
    return std::nullopt;
  }
  if (auto BaseAddr = Parent->getAddress()) {
    return *BaseAddr + getOffset();
  }
  return std::nullopt;
}

// Present for testing purposes only.
void DataBlock::save(std::ostream& Out) const {
  MessageType Message;
  this->toProtobuf(&Message);
  Message.SerializeToOstream(&Out);
}

// Present for testing purposes only.
DataBlock* DataBlock::load(Context& C, std::istream& In) {
  MessageType Message;
  Message.ParseFromIstream(&In);
  auto DB = DataBlock::fromProtobuf(C, nullptr, Message);
  return DB;
}

File: java/com/grammatech/gtirb/DataBlock.java
/*
 *  Copyright (C) 2020-2021 GrammaTech, Inc.
 *
 *  This code is licensed under the MIT license. See the LICENSE file in the
 *  project root for license terms.
 *
 *  This project is sponsored by the Office of Naval Research, One Liberty
 *  Center, 875 N. Randolph Street, Arlington, VA 22203 under contract #
 *  N68335-17-C-0700.  The content of the information does not necessarily
 *  reflect the position or policy of the Government and no official
 *  endorsement should be inferred.
 *
 */

package com.grammatech.gtirb;

import com.google.protobuf.ByteString;
import com.grammatech.gtirb.proto.ByteIntervalOuterClass;
import com.grammatech.gtirb.proto.DataBlockOuterClass;

/**
 * DataBlock represents a data object, possibly symbolic.
 */
public class DataBlock extends ByteBlock {

    /**
     * Class constructor for a {@link DataBlock} from a protobuf DataBlock.
     * @param  protoBlock  The DataBlock as serialized into a protocol buffer.
     */
    private DataBlock(ByteString protoUuid,
                      ByteIntervalOuterClass.Block protoBlock, long size) {
        super(protoUuid, protoBlock, size);
    }

    /**
     * Class constructor for a {@link DataBlock}.
     */
    public DataBlock(long size, long offset) { super(size, offset); }

    /**
     * De-serialize a {@link DataBlock} from a protobuf Block.
     *
     * @return An initialized DataBlock.
     */
    static DataBlock fromProtobuf(ByteIntervalOuterClass.Block protoBlock) {
        // Avoid using protoBlock.hasData() for compatibility with older
        // protobuf
        if (protoBlock.getValueCase() !=
            ByteIntervalOuterClass.Block.ValueCase.DATA) {
            return null;
        }
        DataBlockOuterClass.DataBlock protoDataBlock = protoBlock.getData();
        return new DataBlock(protoDataBlock.getUuid(), protoBlock,
                             protoDataBlock.getSize());
    }

    /**
     * Serialize this DataBlock into a protobuf.
     *
     * @return Block protocol buffer containing this DataBlock.
     */
    @Override
    ByteIntervalOuterClass.Block.Builder toProtobuf() {
        // The protoBlock is in ByteInterval outer class, and it gets a data
        // block added to it with the setData() method. So first create the
        // protoBlock, then create the protoDataBlock and add it to the
        // protoBlock.
        ByteIntervalOuterClass.Block.Builder protoBlock =
            ByteIntervalOuterClass.Block.newBuilder();

        DataBlockOuterClass.DataBlock.Builder protoDataBlock =
            DataBlockOuterClass.DataBlock.newBuilder();
        protoDataBlock.setUuid(Util.uuidToByteString(this.getUuid()));
        protoDataBlock.setSize(this.getSize());
        protoBlock.setOffset(this.getOffset());
        protoBlock.setData(protoDataBlock);
        return protoBlock;
    }
}
